lxd/storage_volumes.go

```diff
@@ -238,51 +247,45 @@ func storagePoolVolumesGet(d *Daemon, r *http.Request) response.Response {
 		return response.SmartError(err)
 	}
 
-	for _, volume := range imageVolumes {
-		if shared.StringInSlice(volume.Name, projectImages) {
+	var volumes []*api.StorageVolume
+
+	for _, projectVolumes := range projectsVolumes {
+		for _, volume := range projectVolumes {
+			// Filter out image volumes that are not used by this project.
+			if volume.Type == db.StoragePoolVolumeTypeNameImage && !shared.StringInSlice(volume.Name, projectImages) {
+				continue
+			}
+
 			volumes = append(volumes, volume)
 		}
 	}
 
 	volumes = filterVolumes(volumes, clauses)
 
-	resultString := []string{}
-	for _, volume := range volumes {
+	var urls []string
+	for _, vol := range volumes {
 		if !recursion {
-			volName, snapName, ok := shared.InstanceGetParentAndSnapshotName(volume.Name)
-			if ok {
-				if projectName == project.Default {
-					resultString = append(resultString,
-						fmt.Sprintf("/%s/storage-pools/%s/volumes/%s/%s/snapshots/%s",
-							version.APIVersion, poolName, volume.Type, volName, snapName))
-				} else {
-					resultString = append(resultString,
-						fmt.Sprintf("/%s/storage-pools/%s/volumes/%s/%s/snapshots/%s?project=%s",
-							version.APIVersion, poolName, volume.Type, volName, snapName, projectName))
-				}
+			u := api.NewURL()
+			volName, snapName, isSnap := shared.InstanceGetParentAndSnapshotName(vol.Name)
+			if isSnap {
+				u = u.Path(version.APIVersion, "storage-pools", poolName, "volumes", vol.Type, volName, "snapshots", snapName).Project(projectName)
 			} else {
-				if projectName == project.Default {
-					resultString = append(resultString,
-						fmt.Sprintf("/%s/storage-pools/%s/volumes/%s/%s",
-							version.APIVersion, poolName, volume.Type, volume.Name))
-				} else {
-					resultString = append(resultString,
-						fmt.Sprintf("/%s/storage-pools/%s/volumes/%s/%s?project=%s",
-							version.APIVersion, poolName, volume.Type, volume.Name, projectName))
-				}
+				u = u.Path(version.APIVersion, "storage-pools", poolName, "volumes", vol.Type, volName).Project(projectName)
 			}
+
+			urls = append(urls, u.String())
 		} else {
-			volumeUsedBy, err := storagePoolVolumeUsedByGet(d.State(), projectName, poolName, volume)
+			volumeUsedBy, err := storagePoolVolumeUsedByGet(d.State(), projectName, poolName, vol)
 			if err != nil {
 				return response.InternalError(err)
 			}
 
-			volume.UsedBy = project.FilterUsedBy(r, volumeUsedBy)
+			vol.UsedBy = project.FilterUsedBy(r, volumeUsedBy)
 		}
 	}
 
 	if !recursion {
-		return response.SyncResponse(true, resultString)
+		return response.SyncResponse(true, urls)
```

Comment:
@stgraber in general do we expect a specific ordering of URLs (or recursive entities) in API output?
The use of maps returned from `GetStoragePoolVolumes`  means the ordering is now random.

Reply:
I'd prefer if we had it be alphabetical. I don't think it's a strict requirement but it may cause some issues with clients expecting it to be at least constant between queries.

Reply:
Thanks, fixed.